### DISCUSSION TITLE
Request for export as an epub

### DISCUSSION THREAD

**Discussion by Jwyat30** [vkbo/novelWriter#1947] [Ideas]
I would for the ability to export EPUBs. I believe that you'll need to incorporate a module or script that can handle EPUB creation and integrate it with novelWriter's existing export functionality. This can be done by:

    Install an EPUB library: First, install a Python library that can create EPUB files. One popular library is ebooklib.

    bash

pip install EbookLib

Create a function to generate EPUB: Write a function that takes the content of your novel from novelWriter and exports it as an EPUB file. Below is an example function using ebooklib:

python

    from ebooklib import epub

    def export_to_epub(novel_content, file_path):
        book = epub.EpubBook()

        # Set metadata
        book.set_identifier('id123456')
        book.set_title('My Novel')
        book.set_language('en')
        
        # Add author
        book.add_author('Amelia Davenport')

        # Add content
        for chapter in novel_content:
            c = epub.EpubHtml(title=chapter['title'], file_name=f'chap_{chapter["id"]}.xhtml', lang='en')
            c.content = chapter['content']
            book.add_item(c)
            book.toc.append(c)

        # Add default NCX and Nav files
        book.add_item(epub.EpubNcx())
        book.add_item(epub.EpubNav())

        # Define CSS style
        style = 'BODY { color: white; }'
        nav_css = epub.EpubItem(uid="style_nav", file_name="style/nav.css", media_type="text/css", content=style)
        book.add_item(nav_css)

        # Create spine
        book.spine = ['nav'] + [epub.Link(chap['file_name'], chap['title'], chap['id']) for chap in novel_content]

        # Write to the file
        epub.write_epub(file_path, book, {})

    Integrate with novelWriter: Modify novelWriter to include this function as part of its export options. This will likely involve editing the codebase of novelWriter, adding a new menu item or button for "Export as EPUB," and connecting it to the export_to_epub function.

Here’s an example of how you might integrate this function in a hypothetical novelWriter codebase:

python

class NovelWriter:
    # Existing code...

    def export_as_epub(self):
        # Get the content from the novelWriter project
        novel_content = self.get_novel_content()
        
        # Define the output file path
        file_path = 'output/my_novel.epub'
        
        # Call the export_to_epub function
        export_to_epub(novel_content, file_path)
        
        print(f'EPUB file created at {file_path}')
        
    def get_novel_content(self):
        # This method should extract the content from the novelWriter project
        # and return it in the format expected by export_to_epub
        return [
            {'id': 1, 'title': 'Chapter 1', 'content': '<h1>Chapter 1</h1><p>This is the first chapter.</p>'},
            {'id': 2, 'title': 'Chapter 2', 'content': '<h1>Chapter 2</h1><p>This is the second chapter.</p>'},
            # Additional chapters...
        ]
        
    # Existing code...

# Example usage
novel_writer = NovelWriter()
novel_writer.export_as_epub()

In this example:

    The export_to_epub function generates the EPUB file.
    The NovelWriter class has a new export_as_epub method that calls export_to_epub with the novel content.
    The get_novel_content method is a placeholder for extracting the content from the novelWriter project.

You would need to adapt this code to fit the actual structure and data management of novelWriter.

**Comment by vkbo** [comment]
I may add EPUB support, but it won't be through an external Python package. I want to avoid having to deal with dependencies outside the bare minimum. At present, the recommended way to generate EPUB is to do it from either Markdown or HTML using Calibre or Pandoc. I'm not entirely sure about the process as I don't really work with EPUB. It is also a little out of scope for novelWriter, since the purpose of it is to generate a manuscript, and EPUB is more of a publishing format. Although I see there are use cases for EPUB in between these boundaries.

In any case, please create a feature request for it, and I will consider it. There's no need for the example code as none of it is relevant to how novelWriter handles output formats or how they are integrated into the UI. The correct way is to subclass the `Tokenizer` class and add the specific formatting code for EPUB there, and add a new build format in Manuscript Build.

**Comment by peter88213** [comment]
May I remark that LibreOffice can export Epub out of the box?

**Reply by vkbo** [reply]
Good point.
